File: extensions/spark/setup/djl_spark/task/audio/speech_recognizer.py
#!/usr/bin/env python
#
# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file
# except in compliance with the License. A copy of the License is located at
#
# http://aws.amazon.com/apache2.0/
#
# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS"
# BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for
# the specific language governing permissions and limitations under the License.

from pyspark import SparkContext
from pyspark.sql import DataFrame
from typing import Optional


class SpeechRecognizer:

    def __init__(self,
                 input_col: str,
                 output_col: str,
                 model_url: str,
                 engine: Optional[str] = None,
                 batch_size: Optional[int] = None,
                 translator_factory=None,
                 batchifier: Optional[str] = None,
                 channels: Optional[int] = None,
                 sample_rate: Optional[int] = None,
                 sample_format: Optional[int] = None):
        """
        Initializes the SpeechRecognizer.

        :param input_col: The input column
        :param output_col: The output column
        :param model_url: The model URL
        :param engine (optional): The engine
        :param batch_size (optional): The batch size
        :param translator_factory (optional): The translator factory.
                                              Default is SpeechRecognitionTranslatorFactory.
        :param batchifier (optional): The batchifier. Valid values include "none" (default),
                                      "stack", and "padding".
        :param channels (optional): The number of channels
        :param sample_rate (optional): The audio sample rate
        :param sample_format (optional): The audio sample format
        """
        self.input_col = input_col
        self.output_col = output_col
        self.model_url = model_url
        self.engine = engine
        self.batch_size = batch_size
        self.translator_factory = translator_factory
        self.batchifier = batchifier
        self.channels = channels
        self.sample_rate = sample_rate
        self.sample_format = sample_format

    def recognize(self, dataset):
        """
        Performs speech recognition on the provided dataset.

        :param dataset: input dataset
        :return: output dataset
        """
        sc = SparkContext._active_spark_context
        recognizer = sc._jvm.ai.djl.spark.task.audio.SpeechRecognizer() \
            .setInputCol(self.input_col) \
            .setOutputCol(self.output_col) \
            .setModelUrl(self.model_url)
        if self.engine is not None:
            recognizer = recognizer.setEngine(self.engine)
        if self.batch_size is not None:
            recognizer = recognizer.setBatchSize(self.batch_size)
        if self.translator_factory is not None:
            recognizer = recognizer.setTranslatorFactory(
                self.translator_factory)
        if self.batchifier is not None:
            recognizer = recognizer.setBatchifier(self.batchifier)
        return DataFrame(recognizer.recognize(dataset._jdf),
                         dataset.sparkSession)

File: extension/spark/src/main/scala/ai/djl/spark/task/audio/SpeechRecognizer.scala
/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 * with the License. A copy of the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
 * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */
package ai.djl.spark.task.audio

import ai.djl.modality.audio.AudioFactory
import ai.djl.modality.audio.translator.SpeechRecognitionTranslatorFactory
import org.apache.spark.ml.param.IntParam
import org.apache.spark.ml.param.shared.{HasInputCol, HasOutputCol}
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.sql.{DataFrame, Dataset, Row}
import org.apache.spark.sql.types.{BinaryType, StringType, StructField, StructType}

import java.io.ByteArrayInputStream
import scala.jdk.CollectionConverters.{collectionAsScalaIterableConverter, seqAsJavaListConverter}

/**
 * SpeechRecognizer performs speech recognition on audio.
 *
 * @param uid An immutable unique ID for the object and its derivatives.
 */
class SpeechRecognizer(override val uid: String) extends BaseAudioPredictor[String]
  with HasInputCol with HasOutputCol {

  def this() = this(Identifiable.randomUID("SpeechRecognizer"))

  final val channels = new IntParam(this, "channels", "The number of channels")
  final val sampleRate = new IntParam(this, "sampleRate", "The audio sample rate")
  final val sampleFormat = new IntParam(this, "sampleFormat", "The audio sample format")

  protected var inputColIndex: Int = _

  /**
   * Sets the inputCol parameter.
   *
   * @param value the value of the parameter
   */
  def setInputCol(value: String): this.type = set(inputCol, value)

  /**
   * Sets the outputCol parameter.
   *
   * @param value the value of the parameter
   */
  def setOutputCol(value: String): this.type = set(outputCol, value)

  /**
   * Sets the channels parameter.
   *
   * @param value the value of the parameter
   */
  def setChannels(value: Int): this.type = set(channels, value)

  /**
   * Sets the sampleRate parameter.
   *
   * @param value the value of the parameter
   */
  def setSampleRate(value: Int): this.type = set(sampleRate, value)

  /**
   * Sets the sampleFormat parameter.
   *
   * @param value the value of the parameter
   */
  def setSampleFormat(value: Int): this.type = set(sampleFormat, value)

  setDefault(outputClass, classOf[String])
  setDefault(translatorFactory, new SpeechRecognitionTranslatorFactory())

  /**
   * Performs speech recognition on the provided dataset.
   *
   * @param dataset input dataset
   * @return output dataset
   */
  def recognize(dataset: Dataset[_]): DataFrame = {
    transform(dataset)
  }

  /** @inheritdoc */
  override def transform(dataset: Dataset[_]): DataFrame = {
    inputColIndex = dataset.schema.fieldIndex($(inputCol))
    super.transform(dataset)
  }

  /** @inheritdoc */
  override protected def transformRows(iter: Iterator[Row]): Iterator[Row] = {
    val predictor = model.newPredictor()
    val audioFactory = AudioFactory.newInstance()
    if (isDefined(channels)) {
      audioFactory.setChannels($(channels))
    }
    if (isDefined(sampleRate)) {
      audioFactory.setSampleRate($(sampleRate))
    }
    if (isDefined(sampleFormat)) {
      audioFactory.setSampleFormat($(sampleFormat))
    }
    iter.grouped($(batchSize)).flatMap { batch =>
      // Read inputs
      val inputs = batch.map { row =>
        val data = row.getAs[Array[Byte]](inputColIndex)
        val is = new ByteArrayInputStream(data)
        try {
          audioFactory.fromInputStream(is)
        } finally {
          is.close()
        }
      }.asJava

      // Batch predict
      val output = predictor.batchPredict(inputs).asScala
      batch.zip(output).map { case (row, out) =>
        Row.fromSeq(row.toSeq :+ out)
      }
    }
  }

  /** @inheritdoc */
  override protected def validateInputType(schema: StructType): Unit = {
    validateType(schema($(inputCol)), BinaryType)
  }

  /** @inheritdoc */
  override def transformSchema(schema: StructType): StructType = {
    val outputSchema = StructType(schema.fields :+ StructField($(outputCol), StringType))
    outputSchema
  }
}
